Add a @_nonSendable attribute

beccadax opened the pull request:
This PR adds a `@_nonSendable` attribute which synthesizes an unavailable `Sendable` conformance on a type.

Using this attribute from clang is [currently being pitched](https://forums.swift.org/t/pitch-2-staging-in-sendable-checking/52413); some aspects of the attribute's design reflect that pitch, but are not fully implemented yet. In particular, its interaction with explicit `Sendable` conformances hasn't really been developed yet.

As a side effect, using an unavailable `Sendable` conformance in pre-Swift 6-mode code is now either a warning or entirely ignored instead of being an error. This is roughly in line with what would happen if the `Sendable` conformance was not present at all (although the check currently in place is again a rough approximation of the pitch).

`@_nonSendable` is currently *not* printed into module interfaces. Instead, we print the synthesized extension, which is more backwards-compatible.

(I've also slipped in a small cleanup of basic attribute argument parsing.)

beccadax committed:
[NFC] Refactor attribute option parsing

beccadax committed:
[NFC] Allow extensions in SynthesizedFileUnits

beccadax committed:
Print @_nonSendable extension in module interfaces

…instead of printing the attribute itself. This allows uses of @_nonSendable to back-deploy.

beccadax commented:
@swift-ci please smoke test

beccadax commented:
Fixup is adding comments and deleting an unused diagnostic—no substantial changes.

@swift-ci please smoke test

slavapestov commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -3693,22 +3693,63 @@ NormalProtocolConformance *GetImplicitSendableRequest::evaluate(
+      auto inherits = ctx.AllocateCopy(makeArrayRef(
+          InheritedEntry(TypeLoc::withoutLoc(proto->getDeclaredInterfaceType()),
+                         /*isUnchecked*/true)));
+      auto extension = ExtensionDecl::create(ctx, attrMakingUnavailable->AtLoc,
```
You might be able to hook this into a lower level in `ConformanceLookupTable` and create the conformance there, instead of actually synthesizing an ExtensionDecl.

beccadax replied:
For now, I think this way is actually better because we can print the extension into module interfaces through normal paths. If we eventually introduce a new syntax for applying availability directly to a conformance, we should probably drop this and switch over to using that instead.

beccadax committed:
Add @_nonSendable decl attribute

This attribute creates an unavailable extension with a `Sendable` conformance so that the type is explicity marked as not being `Sendable`.

We also fully suppress diagnostics about unavailable Sendable conformances in Swift 5 mode code. (This is not fully developed yet—it should return to being a warning in concurrent contexts.)

The behavior when a @_nonSendable and a Sendable conformance are both on the same type is also not right yet.
